run_tests.py

```diff
@@ -53,117 +59,185 @@ def modified_environment(**kwargs):
     os.environ.update(old_env)
 
 
-def run_test(prefix, device, ioc_launcher, lewis_launcher):
+def run_test(prefix, test_module, device_launchers):
     """
-    Runs the tests for the specified IOC.
+    Runs the tests for the specified set of devices.
 
     :param prefix: the instrument prefix
-    :param device: the name of the IOC type
-    :param ioc_launcher: the ioc launcher
-    :param lewis_launcher: the lewis simulator to use; To not use use the LewisNone object
+    :param test_module: the test module
+    :param device_launchers: context manager that launches the necessary iocs and associated emulators
     """
     # Define an environment variable with the prefix in it
     # This can then be accessed elsewhere
     os.environ["testing_prefix"] = prefix
 
-    # Load the device's test module and get the class
-    m = load_module('tests.%s' % device.lower())
+    # Need to set epics address list to local broadcast otherwise channel access won't work
+    settings = {
+        'EPICS_CA_ADDR_LIST': "127.255.255.255"
+    }
+
+    with modified_environment(**settings), device_launchers:
+
+        runner = xmlrunner.XMLTestRunner(output='test-reports')
+
+        test_classes = [getattr(test_module, s) for s in dir(test_module) if s.endswith("Tests")]
+
+        if len(test_classes) < 1:
+            raise ValueError("No test suites found in {}".format(test_module.__name__))
 
-    test_class = getattr(m, "%sTests" % device.capitalize())
+        test_results = []
+        for test_class in test_classes:
+            print("Running tests in {}".format(test_class.__name__))
+            test_suite = unittest.TestLoader().loadTestsFromTestCase(test_class)
+            test_results.append(runner.run(test_suite).wasSuccessful())
+        return all(result is True for result in test_results)
+
+
+def make_device_launchers_from_module(test_module, recsim):
+    """
+    Returns a list of device launchers for the given test module
+    :param test_module: module containing IOC tests
+    :param recsim: True to run in recsim. False to run in devsim.
+    :return: list of device launchers (context managers which launch ioc + emulator pairs)
+    """
 
-    port = str(get_free_ports(1)[0])
-    lewis_launcher.port = port
-    ioc_launcher.port = port
     try:
-        ioc_launcher.macros = m.MACROS
+        iocs = test_module.IOCS
     except AttributeError:
-        ioc_launcher.macros = {}
+        raise AttributeError("Expected module '{}' to contain an IOCS attribute".format(test_module.__name__))
+
+    if len(iocs) < 1:
+        raise ValueError("Need at least one IOC to launch")
+
+    print("Testing module {} in {} mode.".format(test_module.__name__, "recsim" if recsim else "devsim"))
+
+    device_launchers = []
+    for ioc in iocs:
+
+        free_port = str(get_free_ports(1)[0])
+        macros = ioc["macros"]
+        macros['EMULATOR_PORT'] = free_port
+
+        ioc_launcher = IocLauncher(device=ioc["name"],
```

Comment:
Can you put in a better error message for when you don't specify the required keys. Just getting an `KeyError: name` for the test you've just created isn't very obvious.